votor/src/consensus_pool/vote_pool.rs

```diff
@@ -1,129 +1,304 @@
+    std::collections::{btree_map::Entry, BTreeMap},
+    thiserror::Error,
 };
 
-/// There are two types of vote pools:
-/// - SimpleVotePool: Tracks all votes of a specfic vote type made by validators for some slot N, but only one vote per block.
-/// - DuplicateBlockVotePool: Tracks all votes of a specfic vote type made by validators for some slot N,
-///   but allows votes for different blocks by the same validator. Only relevant for VotePool's that are of type
-///   Notarization or NotarizationFallback
-pub(super) enum VotePool {
-    SimpleVotePool(SimpleVotePool),
-    DuplicateBlockVotePool(DuplicateBlockVotePool),
+#[derive(Debug, PartialEq, Eq, Error)]
+pub(crate) enum AddVoteError {
+    #[error("duplicate vote")]
+    Duplicate,
+    #[error("slashable behavior")]
+    Slash,
 }
 
-#[derive(Default)]
-pub(super) struct SimpleVotePool {
-    votes: Vec<VoteMessage>,
-    total_stake: Stake,
-    prev_voted_validators: BTreeSet<Pubkey>,
+/// Container to store per slot votes.
+struct Votes {
+    /// The slot this instance of Votes is responsible for.
+    slot: Slot,
+    /// Skip votes are stored in map indexed by validator.
+    skip: BTreeMap<Pubkey, VoteMessage>,
+    /// Skip fallback votes are stored in map indexed by validator.
+    skip_fallback: BTreeMap<Pubkey, VoteMessage>,
+    /// Finalize votes are stored in map indexed by validator.
+    finalize: BTreeMap<Pubkey, VoteMessage>,
+    /// Notar votes are stored in map indexed by validator.
+    notar: BTreeMap<Pubkey, VoteMessage>,
+    /// A validator can vote notar fallback on upto 3 blocks.
+    ///
+    /// Per validator, we store a map of which block ids the validator has voted notar fallback on.
+    notar_fallback: BTreeMap<Pubkey, BTreeMap<Hash, VoteMessage>>,
 }
 
-impl SimpleVotePool {
-    pub(super) fn add_vote(
-        &mut self,
-        validator_vote_key: Pubkey,
-        validator_stake: Stake,
-        vote: VoteMessage,
-    ) -> Option<Stake> {
-        if !self.prev_voted_validators.insert(validator_vote_key) {
-            return None;
+impl Votes {
+    fn new(slot: Slot) -> Self {
+        Self {
+            slot,
+            skip: BTreeMap::default(),
+            skip_fallback: BTreeMap::default(),
+            finalize: BTreeMap::default(),
+            notar: BTreeMap::default(),
+            notar_fallback: BTreeMap::default(),
         }
-        self.votes.push(vote);
-        self.total_stake = self.total_stake.saturating_add(validator_stake);
-        Some(self.total_stake)
     }
 
-    pub(super) fn votes(&self) -> &[VoteMessage] {
-        &self.votes
-    }
-
-    pub(super) fn total_stake(&self) -> Stake {
-        self.total_stake
+    /// Adds votes.
+    ///
+    /// Checks for different types of slashable behavior and duplicate votes returning appropriate errors.
+    fn add_vote(&mut self, voter: Pubkey, vote: VoteMessage) -> Result<(), AddVoteError> {
+        assert_eq!(self.slot, vote.vote.slot());
```

Comment:
Please make it a habit to rethink every assert_eq, even if cosmic ray caused the slot to differ here, the code should drop these votes and vote pools on other slots might still work. This is production code which is supposed to survive the darkest time.

Only panic when you are about to make damage even if you log error and exit.

Reply:
converted all to debug asserts.